### DISCUSSION TITLE
How to get selected element's full qualified name and it's documentation from source code

### DISCUSSION THREAD

**Discussion by travkin79** [eclipse/lsp4e#911] [Q&A] [answered]
## My Goal
I'd like to get (or calculate) the **full qualified name** of the element under caret (or of the selected element) in an LSP-based editor (in my case LSP-based C++ editor), which I hope to get from the language server. Using the qualified name, I'd like to look up the element's documentation from a separate model (which complements the source code). If such a documentation does not exist, I'd like to get the **element's documentation** from the source code, which I need to find with the language server's help.

## Question
Does anyone has some advice how to do that with LSP4E (or CDT LSP)?

I saw, the LSP protocol provides features like [document symbols](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#textDocument_documentSymbol) and [semantic tokens](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#textDocument_semanticTokens), but both do not provide qualified names and the document symbols do not cover comments.

Besides, there are some [LSP protocol extensions for clangd](https://clangd.llvm.org/extensions), e.g. [symbol info request](https://clangd.llvm.org/extensions#symbol-info-request) and the [abstract syntax tree](https://clangd.llvm.org/extensions#ast). These seem to be the most promising language server interfaces that I could use. It seems, the CDT LSP team [started adding support for these clangd LSP extensions](https://github.com/eclipse-cdt/cdt-lsp/blob/d4bca9e8bf25d427a6a113cf19e222181cfc1218/bundles/org.eclipse.cdt.lsp/src/org/eclipse/cdt/lsp/services/ClangdLanguageServer.java#L24).

Are there any opinions on that or any advice?

**Comment by mickaelistria** [comment]
> Using the qualified name, I'd like to look up the element's documentation from a separate model (which complements the source code)

You could create a LS dedicated to this alternative model, and that would contribute a `hover`. Then, if this extra LS is plugged in, LSP4E would show the inputs from both sources.

**Comment by travkin79** [comment]
I'm not sure, how or if that can work. In my case, the model complementing a C++ file is in a separate file (e.g. an EMF-based model) and contains details for multiple C++ Eclipse projects. Doesn't the _hover_ feature require that model file to be open in a text editor having the cursor at the position corresponding to the element under cursor in the C++ editor?

I originally thought of receiving an ID or URI from the language server and using that ID / URI to look up the documentation in the separate model file or picking up the corresponding comment in C++ source code using the AST.

**Comment by travkin79** [comment]
It turns out, one way of getting an element's (document symbol's) documentation is using the [textDocument/hover](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#textDocument_hover) request type. At least with C++ and clangd, I was able to receive the documentation (and some other details) in a Markdown source code.

But I didn't find a promising way to get an element's qualified name from the language server. I think, I'll have to calculate it myself.

**Reply by mickaelistria** [reply, marked as answer]
> element's qualified name from the language server

LSP doesn't have the concept of a "qualified name", it's all UI/text operations. LSP4E will never be able to provide a standard way to get something that has no definition in LSP. So either you can resolve the qualified name from an existing standard operation (eg hover), or you'll need the LS to provide a custom operation to get the data you want; but in any case it's a LS-specific workflow that LSP4E itself won't be able to provide.

**Reply by travkin79** [reply]
I didn't consider that qualified names might be language-specific. I guess, they are. In that case, it doesn't make sense to provide a generic API for them. Thank you for that hint.

For clangd the LSP extension [textDocument/symbolInfo](https://clangd.llvm.org/extensions#symbol-info-request) seems to offer the details that I need to calculate the qualified name (the symbol name and its parent's qualified name), but that request type is not LSP standard and is clangd-specific.
